lgc/patch/Patch.cpp

```diff
@@ -142,61 +155,39 @@ void Patch::addPasses(PipelineState *pipelineState, lgc::PassManager &passMgr, b
   passMgr.addPass(PatchEntryPointMutate());
   passMgr.addPass(PatchInitializeWorkgroupMemory());
   passMgr.addPass(PatchInOutImportExport());
+  passMgr.addPass(PatchPreparePipelineAbi());
 
-  // Prior to general optimization, do function inlining and dead function removal to remove helper functions that
-  // were introduced during lowering (e.g. streamout stores).
-  passMgr.addPass(AlwaysInlinerPass());
-  passMgr.addPass(GlobalDCEPass());
-
-  passMgr.addPass(createModuleToFunctionPassAdaptor(createFunctionToLoopPassAdaptor(PatchLoopMetadata())));
+  passMgr.addPass(createModuleToFunctionPassAdaptor(PatchBufferOp()));
 
+  // Post-lowering "cleanup" optimizations.
   if (patchTimer) {
     LgcContext::createAndAddStartStopTimer(passMgr, patchTimer, false);
     LgcContext::createAndAddStartStopTimer(passMgr, optTimer, true);
   }
 
-  addOptimizationPasses(passMgr, optLevel);
-
-  if (patchTimer) {
-    LgcContext::createAndAddStartStopTimer(passMgr, optTimer, false);
-    LgcContext::createAndAddStartStopTimer(passMgr, patchTimer, true);
-  }
-
-  // Second part of lowering to "AMDGCN-style"
-  passMgr.addPass(PatchPreparePipelineAbi());
+  // Eliminate helper functions that were introduced during lowering (e.g. for streamout store, and for NGG) and
+  // finalize shader merging (gfx9+).
+  passMgr.addPass(AlwaysInlinerPass());
+  passMgr.addPass(GlobalDCEPass());
 
+  FunctionPassManager fpm;
   const bool canUseNgg = pipelineState->isGraphics() && pipelineState->getTargetInfo().getGfxIpVersion().major == 10 &&
                          (pipelineState->getOptions().nggFlags & NggFlagDisable) == 0;
   if (canUseNgg) {
-    if (patchTimer) {
-      LgcContext::createAndAddStartStopTimer(passMgr, patchTimer, false);
-      LgcContext::createAndAddStartStopTimer(passMgr, optTimer, true);
-    }
-
-    // Extra optimizations after NGG primitive shader creation
-    passMgr.addPass(AlwaysInlinerPass());
-    passMgr.addPass(GlobalDCEPass());
-    FunctionPassManager fpm;
     fpm.addPass(PromotePass());
     fpm.addPass(ADCEPass());
-    fpm.addPass(PatchBufferOp());
-    fpm.addPass(SinkingPass());
-    fpm.addPass(InstCombinePass());
-    fpm.addPass(SimplifyCFGPass());
-    passMgr.addPass(createModuleToFunctionPassAdaptor(std::move(fpm)));
+  }
+  fpm.addPass(InstCombinePass());
+  fpm.addPass(SimplifyCFGPass());
+  passMgr.addPass(createModuleToFunctionPassAdaptor(std::move(fpm)));
 
-    if (patchTimer) {
-      LgcContext::createAndAddStartStopTimer(passMgr, optTimer, false);
-      LgcContext::createAndAddStartStopTimer(passMgr, patchTimer, true);
-    }
-  } else {
-    FunctionPassManager fpm;
-    fpm.addPass(PatchBufferOp());
-    fpm.addPass(SinkingPass());
```

Comment:
I know it's a draft, but I noticed that the patch drops code sinking pass - probably a merge issue.

Reply:
Thanks; and yeah, that must have gotten lost in a rebase.

I noticed some other codegen weirdness in game shaders that I want to investigate first, but I don't know if I'll get to that before my vacation.